src/nnbench/types.py

```diff
@@ -110,22 +126,90 @@ class LocalArtifactLoader(ArtifactLoader):
     Parameters
     ----------
     path : str | os.PathLike[str]
-        The file system pathto the artifact.
+        The file system path to the artifact.
     """
 
-    def __init__(self, path: str | os.PathLike[str]):
+    def __init__(self, path: str | os.PathLike[str]) -> None:
         self._path = path
 
-    def load(self):
+    def load(self) -> Path:
         """
         Returns the path to the artifact on the local file system.
         """
-        return self._path
+        return Path(self._path).resolve()
 
 
-class S3ArtifactLoader(ArtifactLoader):
-    # TODO: Implement this and other common ArtifactLoders here or in a util
-    pass
+class FilePathArtifactLoader(ArtifactLoader):
+    """
+    ArtifactLoader for loading artifacts using fsspec-supported file systems.
+
+    This allows for loading from various file systems like local, S3, GCS, etc.,
+    by using a unified API provided by fsspec.
+
+    Parameters
+    ----------
+    path : str | os.PathLike[str]
+        The path to the artifact, which can include a protocol specifier (like 's3://') for remote access.
+    destination : str | os.PathLike[str] | None
+        The local directory to which remote artifacts will be downloaded. If provided, the model data will be persisted. Otherwise, local artifacts are cleaned.
+    storage_options : dict[str, Any] | None
+        Storage options for remote storage.
+    """
+
+    def __init__(
+        self,
+        path: str | os.PathLike[str],
+        destination: str | os.PathLike[str] | None = None,
+        storage_options: dict[str, Any] | None = None,
+    ) -> None:
+        self.source_path = str(path)
+        if destination:
+            self.target_path = str(Path(destination).resolve())
+            delete = False
+        else:
+            self.target_path = str(Path(mkdtemp()).resolve())
+            delete = True
+        self._finalizer = weakref.finalize(self, self._cleanup, delete=delete)
+        self.storage_options = storage_options or {}
+
+    def _get_protocol(self, path: str) -> str:
+        """Lazy loading of fsspec.utils.get_protocol"""
+        try:
+            from fsspec.utils import get_protocol
+
+            return get_protocol(path)
+        except ImportError:
+            raise ImportError(
+                "class {self.__class__.__name__} requires `fsspec` to be installed. You can install it by running `python -m pip install --upgrade fsspec`"
+            )
```

Comment:
Two things here:

1) The single use in `load()` below does not warrant extracting these into their own APIs - please just import the fsspec versions both inline in `load()`, and scrap the private class methods.
2) You can factor out the import error with a simple import guard at the top of the file, e.g. like so:

```python
# this towards the top, after the imports.

try:
    import fsspec
    HAS_FSSPEC = True
except ImportError:
    HAS_FSSPEC = False

...
# later, in FilePathArtifactLoader.__init__():
...
if not HAS_FSSPEC:
    raise ModuleNotFoundError("class blah requires `fsspec`...")
```